### PR TITLE
Introducing modules for the ts client

### PR DESCRIPTION
**Pull request by moozzyk**

### CODE REVIEW

**Comment by davidfowl** [comment]
Is this the first pass? Is there any way to do this without using requirejs?

**Comment by moozzyk** [comment]
Yes this is the first pass. I wanted to give it a spin and see how it works and iterate on this. This should also unblock adding node based tests. 
Re requirejs - you need a loader to be able to load amd modules. One idea that I have not yet full investigated is using es2015/es6 modules. Looks like it is not yet widely supported by browsers (only Edge supports it) so on other browsers you would still need to use a js library to use it. es2015 modules should be compatible with node but as I said I have not tried it yet. 
We probably could also just have just a js file but we would have to make sure it does not dump everything in the global scope.

**Comment by davidfowl** [comment]
> We probably could also just have just a js file but we would have to make sure it does not dump everything in the global scope.

We need a single js file, we can't require a loader.

**Comment by moozzyk** [comment]
:up::date:

File: samples/ChatSample/project.json [hunk 1]
```diff
@@ -91,7 +91,11 @@
   },
 
   "scripts": {
-    "precompile": [ "dotnet bundle", "tsc --project ../../src/Microsoft.AspNetCore.SignalR.Client.TS/ --outDir wwwroot/lib/signalr-client/ --module amd" ],
+    "precompile": [ "dotnet bundle",
```
Comment (moozzyk) [inline samples/ChatSample/project.json]: This is super ugly. I am going to try to move it to gulp entirely.

File: samples/ChatSample/project.json [hunk 2]
```diff
@@ -91,7 +91,11 @@
   },
 
   "scripts": {
-    "precompile": [ "dotnet bundle", "tsc --project ../../src/Microsoft.AspNetCore.SignalR.Client.TS/ --out wwwroot/js/signalr-client.js" ],
+    "precompile": [ "dotnet bundle",
```
Comment (davidfowl) [inline samples/ChatSample/project.json]: This looks kinda crazy

**Comment by moozzyk** [comment]
:up::date: - moved building js client to gulp

File: test/Microsoft.AspNetCore.SignalR.Test.Server/project.json [hunk 3]
```diff
@@ -39,7 +39,7 @@
   },
 
   "scripts": {
-    "precompile": [ "npm install", "gulp copy-jasmine"],
+    "precompile": [ "npm install", "npm run gulp -- --gulpfile %project:Directory%/gulpfile.js copy-jasmine"],
```
Comment (moozzyk) [inline test/Microsoft.AspNetCore.SignalR.Test.Server/project.json]: @pranavkm @davidfowl @muratg @BrennanConroy 
This fixes the CI.

File: .gitignore [hunk 4]
```diff
@@ -35,7 +35,6 @@ node_modules/
 *.nuget.props
 *.nuget.targets
 autobahnreports/
-signalr-client.js
```
Comment (davidfowl) [inline .gitignore]: Why is this removed from ignore now?

Comment (davidfowl) [inline .gitignore]: Ah I see

File: gulpfile.js [hunk 5]
```diff
@@ -0,0 +1,40 @@
+const gulp = require('gulp');
```
Comment (davidfowl) [inline gulpfile.js]: Why is this in the root?

Comment (davidfowl) [inline gulpfile.js]: Why is this in the root?

File: package.json [hunk 6]
```diff
@@ -6,19 +6,27 @@
   "directories": {
```
Comment (davidfowl) [inline package.json]: Why is this in the root?

File: samples/SocketsSample/.gitignore [hunk 7]
```diff
@@ -0,0 +1 @@
+wwwroot/js
```
Comment (davidfowl) [inline samples/SocketsSample/.gitignore]: Why not globally?

File: samples/SocketsSample/wwwroot/.gitignore [hunk 8]
```diff
@@ -1 +1 @@
-js/
\ No newline at end of file
+lib/signalr-client
```
Comment (davidfowl) [inline samples/SocketsSample/wwwroot/.gitignore]: Why do we have gitignores per sample? We don't do this anywhere else

File: gulpfile.js (hunk 5 above)
Comment (moozzyk) [inline gulpfile.js]: because it's common for multiple sample projects

File: package.json (hunk 6 above)
Comment (moozzyk) [inline package.json]: because it is common to multiple projects. Also avoids having node_modules in each projects that uses node

File: gulpfile.js (hunk 5 above)
Comment (davidfowl) [inline gulpfile.js]: Having it in the root looks untidy. Put it in the samples folder.

File: samples/SocketsSample/.gitignore (hunk 7 above)
Comment (davidfowl) [inline samples/SocketsSample/.gitignore]: Can we be consistent about this? I don't like the per project git ignore. Lets just keep the global signalr-client.js include for now...

File: samples/SocketsSample/wwwroot/.gitignore (hunk 8 above)
Comment (moozzyk) [inline samples/SocketsSample/wwwroot/.gitignore]: If I unified the projects and put signalr in the same relative path we could move it to the common .gitignore.

File: gulpfile.js (hunk 5 above)
Comment (moozzyk) [inline gulpfile.js]: but it is not just samples - this is/will be part of the build process. This is how we build the library for browsers (i.e. as a single file).

Comment (moozzyk) [inline gulpfile.js]: If we wanted in the future we will be able to replace gulp with MSBuild...

Comment (moozzyk) [inline gulpfile.js]: Maybe I could move it to a build or tools folder then?

Comment (davidfowl) [inline gulpfile.js]: It really, shouldn't be in the root. It only applies to the project that produces the js file. We're not using gulp for project builds. Move it to the typescript client then.

File: samples/SocketsSample/.gitignore (hunk 7 above)
Comment (moozzyk) [inline samples/SocketsSample/.gitignore]: OK. as per my comment below I can unify where we put signalr (e.g. wwwroot/lib/signalr) and ignore this in the main .gitignore.

**Comment by moozzyk** [comment]
:up::date: Moved gulpfile to the Client.TS folder, unified signalr-client.js location. Removed local .gitignore files in samples.
